### PR TITLE
App Insights control plane

### PR DESCRIPTION
**Pull request by alexeldeib**
---

Introduces commands for managing App Insights components and API keys belonging to those resources. See https://github.com/Azure/azure-cli/issues/5543

### TODO
- [x] test - api key create
- [x] test - api key delete
- [x] test - api key show (single/multiple) 
- [x] test - component create
- [x] test - component update
- [x] test - component delete
- [x] test - component show (single/RG/sub)

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `./scripts/ci/test_static.sh` locally? (`pip install pylint flake8` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [x] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### CODE REVIEW

**Comment by alexeldeib** [comment]
@williexu could you help me decode this CI error? is `--kind` not a valid flag?

```
-  FAIL: unrecognized_help_parameter_rule
    Help-Entry: `monitor app-insights component create` - The following parameter help names are invalid: --kind
    Help-Entry: `monitor app-insights component update` - The following parameter help names are invalid: --kind
```

The associated help text (also tried defining the --kind parameter explicitly here, but maybe I messed it up?):
```
helps['monitor app-insights component create'] = """
    type: command
    short-summary: Create a new Application Insights resource.
    parameters:
      - name: --application-type
        type: string
        short-summary: Type of application being monitored. Possible values include 'web', 'other'. Default value is'web' .
    examples:
      - name: Create a component with kind web and location.
        text: |
          az monitor app-insights component create --app demoApp --location westus2 --kind web -g demoRg --application-type web
"""

helps['monitor app-insights component update'] = """
    type: command
    short-summary: Update properties on an existing Application Insights resource. The primary value which can be updated is kind, which customizes the UI experience.
    examples:
      - name: Update a component with kind web.
        text: |
          az monitor app-insights component update --app demoApp --kind web -g demoRg
"""
```

**Comment by tjprescott** [comment]
@alexeldeib the problem is that you need to specify the long and short option:
```
- name: --kind -k
```

**Comment by alexeldeib** [comment]
@tjprescott could you review this? After those fixes + additions, this looks okay to me.

**Comment by alexeldeib** [comment]
Bump 😃 cc @williexu

File: src/application-insights/azext_applicationinsights/custom.py [hunk 1]
```diff
@@ -31,3 +32,67 @@ def get_metric(cmd, client, application, metric, start_time=None, end_time=None,
...
+    api_key_request = APIKeyRequest(api_key, linked_read_properties, linked_write_properties)
+    return client.create(resource_group_name, application, api_key_request)
+
+
+def show_api_key(client, application, resource_group_name, api_key=None):
+    if api_key is None:
+        return client.list(resource_group_name, application)
+    result = list(filter(lambda result: result.name == api_key, client.list(resource_group_name, application)))
+    if len(result) == 1:
+        return result[0]
```
Comment (yugangw-msft) [inline src/application-insights/azext_applicationinsights/custom.py]: I suggest just return `result` to have consistent return type in array which would help the scripting

File: src/application-insights/azext_applicationinsights/_help.py [hunk 2]
```diff
@@ -21,6 +21,126 @@
...
+    short-summary: Manage an Application Insights component or its subcomponents.
+"""
+
+helps['monitor app-insights component create'] = """
+    type: command
+    short-summary: Create a new Application Insights resource.
+    parameters:
+      - name: --application-type
+        type: string
+        short-summary: Type of application being monitored. Possible values include 'web', 'other'. Default value is'web' .
```
Comment (yugangw-msft) [inline src/application-insights/azext_applicationinsights/_help.py]: Add a space inside `is'web'`?

File: src/application-insights/azext_applicationinsights/_help.py [hunk 3]
```diff
@@ -21,6 +21,126 @@
...
+helps['monitor app-insights component update-tags'] = """
+    type: command
+    short-summary: Update tags on an existing Application Insights resource.
+    examples:
+      - name: Update the tag 'name' to equal 'value'.
+        text: |
+          az monitor app-insights component update-tags --app demoApp --tags name=value -g demoRg
+"""
+
+helps['monitor app-insights component show'] = """
```
Comment (yugangw-msft) [inline src/application-insights/azext_applicationinsights/_help.py]: I suggest add a command of `monitor app-insights component list` for `List components in a resource group or subscription`

File: src/application-insights/azext_applicationinsights/_help.py [hunk 4]
```diff
@@ -21,6 +21,126 @@
...
+      - name: List components in the currently selected subscription.
+        text: |
+          az monitor app-insights component show
+"""
+
+helps['monitor app-insights component delete'] = """
+    type: command
+    short-summary: Create a new Application Insights resource.
+    examples:
+      - name: Create a component with kind web and location.
```
Comment (yugangw-msft) [inline src/application-insights/azext_applicationinsights/_help.py]: probably a copy and paste error from the `create` command?

File: src/application-insights/azext_applicationinsights/_help.py [hunk 5]
```diff
@@ -21,6 +21,126 @@
...
+        text: |
+          az monitor app-insights component delete --app demoApp -g demoRg
+"""
+
+helps['monitor app-insights api-key'] = """
+    type: group
+    short-summary: Operations on API keys associated with an Application Insights component.
+"""
+
+helps['monitor app-insights api-key show'] = """
```
Comment (yugangw-msft) [inline src/application-insights/azext_applicationinsights/_help.py]: Again, please consider have a `monitor app-insights api-key list` on getting a list of key from an app

File: src/application-insights/azext_applicationinsights/_params.py [hunk 6]
```diff
@@ -4,17 +4,36 @@
...
+
+    with self.argument_context('monitor app-insights component create') as c:
+        c.argument('location', arg_type=get_location_type(self.cli_ctx))
+        c.argument('application-type', options_list=['application-type', '--type', '-t'], help="Type of application being monitored. Possible values include: 'web', 'other'. Default value: 'web' .")
+        c.argument('kind', options_list=['--kind', '-k'], help='The kind of application that this component refers to, used to customize UI. This value is a freeform string, values should typically be one of the following: web, ios, other, store, java, phone.')
+        c.argument('tags', tags_type)
+
+    with self.argument_context('monitor app-insights component update') as c:
+        c.argument('location', arg_type=get_location_type(self.cli_ctx))
+        c.argument('application-type', options_list=['application-type', '--type', '-t'], help="Type of application being monitored. Possible values include: 'web', 'other'. Default value: 'web' .")
```
Comment (yugangw-msft) [inline src/application-insights/azext_applicationinsights/_params.py]: I assume you will get `--application-type` even w/o the leading `-`?

File: src/application-insights/azext_applicationinsights/_params.py [hunk 7]
```diff
@@ -4,17 +4,36 @@
...
+        c.argument('kind', options_list=['--kind', '-k'], help='The kind of application that this component refers to, used to customize UI. This value is a freeform string, values should typically be one of the following: web, ios, other, store, java, phone.')
+
+    with self.argument_context('monitor app-insights component update-tags') as c:
+        c.argument('tags', tags_type)
+
+    with self.argument_context('monitor app-insights api-key create') as c:
+        c.argument('api_key', help='The name of the API key to create.')
+        c.argument('read_properties', nargs='+', options_list=['--read-properties'])
```
Comment (yugangw-msft) [inline src/application-insights/azext_applicationinsights/_params.py]: `, options_list=['--read-properties']` is redundant

**Comment by alexeldeib** [comment]
@yugangw-msft fixed all comments. Two test errors look related to AKS. Additionally, now that I have `az monitor api-key list`, i'm getting an error from CI on `no --ids with list commands` (but api-keys are under components). Can I disable this error? 

Otherwise, think this is okay again.

File: src/application-insights/azext_applicationinsights/_help.py (hunk 5 above)
Comment (alexeldeib) [reply src/application-insights/azext_applicationinsights/_help.py]: fixed

File: src/application-insights/azext_applicationinsights/_params.py (hunk 6 above)
Comment (alexeldeib) [reply src/application-insights/azext_applicationinsights/_params.py]: aligned for consistency

File: src/application-insights/azext_applicationinsights/_help.py (hunk 3 above)
Comment (alexeldeib) [reply src/application-insights/azext_applicationinsights/_help.py]: added

**Comment by alexeldeib** [comment]
@yugangw-msft bringing this back after the weekend, barring changes on the `show` with `ids`, I think this is good on my end.

**Comment by yugangw-msft** [comment]
@alexeldeib, you can disable the error. Can you please rebase to get CI green?

**Comment by alexeldeib** [comment]
@yugangw-msft Done. Sorry, missed that I needed rebase.

**Comment by alexeldeib** [comment]
Whoops, messed that up. One sec...

**Comment by alexeldeib** [comment]
@yugangw-msft ok, now the rebase is fixed. If you can let me know the tag to disable, I will add it. I'm checking other commands if they have something similar but haven't found it yet,

**Comment by alexeldeib** [comment]
Okay, actually fixed.

**Pull request merged**